thriftpool/channel.go

```diff
@@ -0,0 +1,120 @@
+package thriftpool
+
+import (
+	"sync/atomic"
+)
+
+type channelPool struct {
+	pool           chan Client
+	opener         ClientOpener
+	numActive      int32
+	minConnections int
+	maxConnections int
+}
+
+// Make sure channelPool implements Pool interface.
+var _ Pool = (*channelPool)(nil)
+
+// NewChannelPool creates a new thrift client pool implemented via channel.
+func NewChannelPool(minConnections, maxConnections int, opener ClientOpener) (Pool, error) {
+	if minConnections > maxConnections {
+		return nil, &ConfigError{
+			MinConnections: minConnections,
+			MaxConnections: maxConnections,
+		}
+	}
+
+	pool := make(chan Client, maxConnections)
+	for i := 0; i < minConnections; i++ {
+		c, err := opener()
+		if err != nil {
+			return nil, err
+		}
+		pool <- c
+	}
+
+	return &channelPool{
+		pool:           pool,
+		opener:         opener,
+		minConnections: minConnections,
+		maxConnections: maxConnections,
+	}, nil
+}
+
+// Get returns a thrift client from the pool.
+func (cp *channelPool) Get() (client Client, err error) {
+	defer func() {
+		if err == nil {
+			atomic.AddInt32(&cp.numActive, 1)
+		}
+	}()
+
+	select {
+	case c := <-cp.pool:
+		if c.IsOpen() {
+			return c, nil
+		}
+	default:
+	}
+
+	if cp.IsExhausted() {
+		err = ErrExhausted
+		return
+	}
+	return cp.opener()
+}
+
+// Release releases a client back to the pool.
+//
+// If the pool is full, the client will be closed instead.
+//
+// Calling Release after Close will cause panic.
+func (cp *channelPool) Release(c Client) error {
+	if c == nil {
+		return nil
+	}
+
+	if !c.IsOpen() {
+		newC, err := cp.opener()
+		if err != nil {
+			return err
+		}
+		c = newC
+	}
+
+	select {
+	case cp.pool <- c:
+		atomic.AddInt32(&cp.numActive, -1)
+		return nil
+	default:
+		// Pool is full, just close it instead.
+		return c.Close()
```

Comment:
🔕 I was trying to think through _how_ this can happen, is it because the following can happen:

1. 2 goroutines call `Get`
2. Both try to get a connection while the pool is empty
3. _Before_ they check `IsExhausted`, a connection is released back into the pool
4. Both goroutines open new connections and return those

Reply:
Note, this can also happen because the pool doesn’t necessarily start full, so there can be 1 slot left and 2 goroutines calling “get” at the same time, so they both can pass the exhausted check and make 2 new connections